src/backend/distributed/stat_counters.c

```diff
@@ -0,0 +1,322 @@
+{
+	Datum values[STAT_COUNTERS_COLUMNS] = { 0 };
+	bool isNulls[STAT_COUNTERS_COLUMNS] = { 0 };
+
+	CitusStatCounters aggregatedStatCounters;
+	MemSet(aggregatedStatCounters, 0, sizeof(CitusStatCounters));
+
+	AggregateStatCountersInto(&aggregatedStatCounters);
+
+	for (int i = 0; i < MAX_STAT_COUNT; i++)
+	{
+		if (aggregatedStatCounters[i] == 0)
+		{
+			continue;
+		}
+
+		values[0] = PointerGetDatum(cstring_to_text(StatMapping[i]));
+		values[1] = Int64GetDatum(aggregatedStatCounters[i]);
+		tuplestore_putvalues(tupleStore, tupleDescriptor, values, isNulls);
+	}
+}
+
+
+/*
+ * AggregateStatCountersInto aggregates the stat counters of all the backends into
+ * the given aggregatedStatCounters.
+ */
+static void
+AggregateStatCountersInto(CitusStatCounters *aggregatedStatCounters)
+{
+	const Oid userId = GetUserId();
+
+	for (int backendIndex = 0; backendIndex < MaxBackends; ++backendIndex)
+	{
+		PGPROC *currentProc = GetPGProcByNumber(backendIndex);
+
+		/*
+		 * We imitate pg_stat_activity such that if a user doesn't have enough
+		 * privileges, then we don't collect the stats for that backend.
+		 */
+		if (UserHasPermissionToViewStatsOf(userId, currentProc->roleId))
+		{
+			for (int statIdx = 0; statIdx < MAX_STAT_COUNT; statIdx++)
+			{
+				(*aggregatedStatCounters)[statIdx] +=
+					pg_atomic_read_u64(&SharedStatCountersArray[backendIndex][statIdx]);
+			}
+		}
+	}
+}
+
+
+/*
+ * ResetStatCounters resets all the stat counters for all the backends.
+ */
+static void
+ResetStatCounters(void)
+{
+	/*
+	 * Some stats might be lost between reading the stats for all the backend processes
+	 * and resetting the stats. However, we are okay with this since we don't want to block
+	 * the client backends that might be incrementing the stats.
+	 *
+	 * Also, we cannot use Memset() to reset the stats as it is not safe to reset the stats
+	 * while other processes are reading them, if any.
+	 */
+	for (int backendIndex = 0; backendIndex < MaxBackends; ++backendIndex)
+	{
+		for (int statIdx = 0; statIdx < MAX_STAT_COUNT; statIdx++)
+		{
+			pg_atomic_write_u64(&SharedStatCountersArray[backendIndex][statIdx], 0);
+		}
+	}
+}
+
+
+/*
+ * SharedStatCountersArrayShmemInit initializes the shared memory used
+ * for keeping track of stat counters across backends.
+ */
+static void
+SharedStatCountersArrayShmemInit(void)
+{
+	StaticAssertExpr(MAX_STAT_INDEX < MAX_STAT_COUNT,
+					 "stat enums should be less than size - bump up MAX_STAT_COUNT");
+
+	/* validate that we have names for the stat counters as well */
+	for (int i = 0; i < MAX_STAT_INDEX; i++)
+	{
+		if (strlen(StatMapping[i]) == 0)
+		{
+			ereport(PANIC, (errmsg("Stat mapping for index %d not found",
+								   i)));
+		}
+	}
+
+	bool alreadyInitialized;
+
+	size_t statCountersBackendArrayShmemSize = StatCountersArrayShmemSize();
+
+	LWLockAcquire(AddinShmemInitLock, LW_EXCLUSIVE);
+
+	SharedStatCountersArray = (CitusAtomicStatCounters *)
```

Comment:
We do not need connection success/failure stats per backend, right? So I assume this is for future proofing if other per-backend stats  are added in future.